cub/cub/agent/agent_topk.cuh

```diff
@@ -0,0 +1,830 @@
+
+  /// The BlockHistogram algorithm to use
+  static constexpr cub::BlockHistogramAlgorithm HISTOGRAM_ALGORITHM = _HISTOGRAM_ALGORITHM;
+
+  /// The BlockScan algorithm to use
+  static constexpr cub::BlockScanAlgorithm SCAN_ALGORITHM = _SCAN_ALGORITHM;
+};
+
+/******************************************************************************
+ * Thread block abstractions
+ ******************************************************************************/
+
+template <typename KeyInT, typename NumItemsT>
+struct alignas(128) Counter
+{
+  // We are processing the values in multiple passes, from most significant to least
+  // significant. In each pass, we keep the length of input (`len`) and the `k` of
+  // current pass, and update them at the end of the pass.
+  NumItemsT k;
+  NumItemsT len;
+
+  //  `previous_len` is the length of input in previous pass. Note that `previous_len`
+  //  rather than `len` is used for the filtering step because filtering is indeed for
+  //  previous pass (see comments before `radix_kernel`).
+  NumItemsT previous_len;
+
+  // We determine the bits of the k_th key inside the mask processed by the pass. The
+  // already known bits are stored in `kth_key_bits`. It's used to discriminate a
+  // element is a result (written to `out`), a candidate for next pass (written to
+  // `out_buf`), or not useful (discarded). The bits that are not yet processed do not
+  // matter for this purpose.
+  typename cub::Traits<KeyInT>::UnsignedBits kth_key_bits;
+
+  // Record how many elements have passed filtering. It's used to determine the position
+  // in the `out_buf` where an element should be written.
+  alignas(128) NumItemsT filter_cnt;
+
+  // For a row inside a batch, we may launch multiple thread blocks. This counter is
+  // used to determine if the current block is the last running block. If so, this block
+  // will execute Scan() and ChooseBucket().
+  alignas(128) unsigned int finished_block_cnt;
+
+  // Record how many elements have been written to the front of `out`. Elements less (if
+  // SELECT_MIN==true) than the k-th key are written from front to back.
+  alignas(128) NumItemsT out_cnt;
+
+  // Record how many elements have been written to the back of `out`. Elements equal to
+  // the k-th key are written from back to front. We need to keep count of them
+  // separately because the number of elements that <= the k-th key might exceed k.
+  alignas(128) NumItemsT out_back_cnt;
+};
+
+/**
+ * @brief Operations for calculating the bin index based on the input
+ */
+template <typename T, bool SELECT_MIN>
+struct ExtractBinOp
+{
+  _CCCL_HOST_DEVICE _CCCL_FORCEINLINE typename cub::Traits<T>::UnsignedBits operator()(T key)
+  {
+    auto bits = reinterpret_cast<typename cub::Traits<T>::UnsignedBits&>(key);
+    bits      = cub::Traits<T>::TwiddleIn(bits);
+    if (!SELECT_MIN)
+    {
+      bits = ~bits;
+    }
+    return bits;
+  }
+};
```

Comment:
Agree with your suggestion. I haven't finished this part.  My concern is about code related to 
https://github.com/ChristinaZ/cccl/blob/christinaz/nondeterministic_AIR_topK/cub/cub/agent/agent_topk.cuh#L527

Previously, we used all the previous bits to compare and find the target element instead of using certain bucket for each pass. I haven't find out the better solution for now. Let me try to find a good solution for it.